Add getting_started to index.rst (#294)
SynBioDex/pySBOL3: docs/getting_started.rst

Ontology terms also take the form of Uniform Resource Identifiers. Many commonly used ontological terms are built-in to pySBOL3 as predefined constants. If an ontology term is not provided as a built-in constant, its URI can often be found by using an ontology browser tool online. `Browse Sequence Ontology terms here <http://www.sequenceontology.org/browser/obob.cgi>`_ and `Systems Biology Ontology terms here <http://www.ebi.ac.uk/sbo/main/tree>`_. While the SBOL specification often recommends particular ontologies and terms to be used for certain attributes, in many cases these are not rigid requirements. The advantage of using a recommended term is that it ensures your data can be interpreted or visualized by other applications that support SBOL. However in many cases an application developer may want to develop their own ontologies to support custom applications within their domain.

The following example illustrates how the URIs for ontology terms can be easily constructed, assuming they are not already part of pySBOL3's built-in ontology constants.

.. code:: python

>     >>> SO_ENGINEERED_FUSION_GENE = sbol3.SO_NS + '0000288'  # Sequence Ontology term

Reviewer: ```suggestion
    >>> SO_ENGINEERED_FUSION_GENE = tyto.SO.engineered_fusion_gene
```

    >>> SO_ENGINEERED_FUSION_GENE
    'https://identifiers.org/SO:0000288'

>     >>> SBO_DNA_REPLICATION = sbol3.SBO_NS + '0000204'  # Systems Biology Ontology term

Reviewer: ```suggestion
    >>> SBO_DNA_REPLICATION = tyto.SBO.DNA_replication
```

Based on our inspection of objects contained in the Document above, we can see that these objects were all created in the namespace ``http://sbolstandard.org/testfiles``. Thus, in order to take advantage of SBOL-compliant URIs, we set an environment variable that configures this namespace as the default.

.. code:: python

    >>> sbol3.set_namespace('http://sbolstandard.org/testfiles')

.. end

Setting the namespace has several advantages. It simplifies object creation and retrieval from Documents. In addition, it serves as a way for a user to claim ownership of new objects. Generally users will want to specify a namespace that corresponds to their organization's web domain.

--------------------------
Creating SBOL Data Objects
--------------------------

Biological designs can be described with SBOL data objects, including both structural and functional features.  

> The principle classes for describing the structure and primary sequence of a design are ``Component``, ``Sequence``, ``SequencFeature``, and ``SubComponent``. 

Reviewer 2: ``Component``, ``Sequence``, and ``Feature``.

Reviewer: Resolved.

> The principle classes for describing the structure and primary sequence of a design are ``Component``, ``Sequence``, ``SequencFeature``, and ``SubComponent``. 

Reviewer: ```suggestion
The principle classes for describing the structure and primary sequence of a design are ``Component``, ``Sequence``, and ``Feature``. 
```

> The principle classes for describing the function of a design are ``Component``, ``SubComponent``, ``Interaction``, and ``Participation``. 

Reviewer 2: ``Feature`` instead of ``SubComponent``

Reviewer: ```suggestion
The principle classes for describing the function of a design are ``Component``, ``Feature``, ``Interaction``, and ``Participation``. 
```

> .. TODO: delete or edit; Is there a pySBOL3 equivalent to "Design, Build, Test, Analysis, Activity, and Plan"

Reviewer: ```suggestion
```

>     Other classes such as Design, Build, Test, Analysis, Activity, and Plan are used for managing workflows.

Reviewer 2: I'd suggest just leaving these out for now.

Reviewer 3: 👍

Reviewer: ```suggestion
```

Constructors can take additional arguments, depending on whether the SBOL class has required attributes. Attributes are required if the specification says they are. In a UML diagram, required attributes are indicated as properties with a cardinality of 1 or more. For example, a ``Component`` (see the UML diagram above) has only one required attribute, ``types``, which specifies one or more molecular types for a component.  Required attributes MUST be specified when calling a constructor. 

The following code creates a protein component (``types`` set to ``SBO_PROTEIN``). 

.. code:: python

    >>> cas9 = sbol3.Component('Cas9', sbol3.SBO_PROTEIN)

.. end

The following code creates a DNA component (``types`` set to ``SBO_DNA``). 

.. code:: python

>     >>> target_promoter = sbol3.Component('target_promoter', sbol3.SBO_DNA)

Reviewer 2: I'd suggest that we also add the role here for a promoter.

Reviewer: ```suggestion
    >>> target_promoter = sbol3.Component('target_promoter', sbol3.SBO_DNA, roles=[sbol3.SO_PROMOTER])
```

to be specific. Ontologies are standardized, machine-readable vocabularies that categorize concepts within a domain of scientific study. The SBOL 3.0 standard unifies many different ontologies into a high-level, object-oriented model.

Ontology terms also take the form of Uniform Resource Identifiers. Many commonly used ontological terms are built-in to pySBOL3 as predefined constants. If an ontology term is not provided as a built-in constant, its URI can often be found by using an ontology browser tool online. `Browse Sequence Ontology terms here <http://www.sequenceontology.org/browser/obob.cgi>`_ and `Systems Biology Ontology terms here <http://www.ebi.ac.uk/sbo/main/tree>`_. While the SBOL specification often recommends particular ontologies and terms to be used for certain attributes, in many cases these are not rigid requirements. The advantage of using a recommended term is that it ensures your data can be interpreted or visualized by other applications that support SBOL. However in many cases an application developer may want to develop their own ontologies to support custom applications within their domain.

> The following example illustrates how the URIs for ontology terms can be easily constructed, assuming they are not already part of pySBOL3's built-in ontology constants.

Reviewer 2: I think this example should use tyto, as is recommended in "using ontology terms"

Reviewer: This is now resolved

    
    >>> ptet_sc = sbol3.SubComponent(ptet)
    >>> circuit.features += [ptet_sc]

.. end

-----------------------------------------
Creating and Editing Reference Properties
-----------------------------------------

Some SBOL objects point to other objects by way of URI references. For example, Components point to their corresponding Sequences by way of a URI reference. These kind of properties correspond to white diamond arrows in UML diagrams, as shown in the figure above. Attributes of this type contain the URI of the related object.

.. code:: python

    >>> gfp = sbol3.Component('GFP', sbol3.SBO_DNA)
    >>> gfp_seq = sbol3.Sequence('GFPSequence', elements='atgnnntaa', encoding=sbol3.IUPAC_DNA_ENCODING)
    >>> gfp.sequences = [ gfp_seq ]
    >>> print(gfp.sequences)
    ['http://sbolstandard.org/testfiles/GFPSequence']

.. end

Note that assigning the ``gfp_seq`` object to the ``gfp.sequences`` actually results in assignment of the object's URI. An equivalent assignment is as follows:

.. code:: python

    >>> gfp.sequences = [ gfp_seq.identity ]
    >>> print(gfp.sequences)

>     ['http://sbolstandard.org/testfiles/GFPSequence']

Reviewer: ```suggestion
    ['http://sbolstandard.org/testfiles/GFPSequence']
    >>> seq2 = gfp.sequences[0].lookup()
    >>> seq2 == gfp_seq
    True
```

Creating and Adding Child Objects
------------------------------------------

Some SBOL objects can be composed into hierarchical parent-child relationships.  
In the specification diagrams, these relationships are indicated by black diamond arrows.  
In the UML diagram above, the black diamond indicates that Components are parents of Features.  
In pySBOL3, properties of this type are created as subcomponents and then added to the appropriate 
list attribute of the parent component. The constructor for the ``SubComponent`` class 
takes a ``Component`` as its only required argument. In this usage, the ``Component`` is "... analogous 
to a blueprint or specification sheet for a biological part..." Whereas the ``SubComponent`` "... represents 
the specific occurrence of a part..." within a larger design 
(`SBOL version 3.0.0 specification document <https://sbolstandard.org/docs/SBOL3.0specification.pdf>`_). 
For example, to add a promoter to a circuit design, first define the promoter and circuit as SBOL 
``Component`` objects, then define a ``SubComponent`` as an instance of the promoter and add that 
``SubComponent`` to the circuit's ``features`` attribute:

.. code:: python

>     >>> ptet = sbol3.Component('pTetR', sbol3.SBO_DNA)

Reviewer: ```suggestion
    >>> ptet = sbol3.Component('pTetR', sbol3.SBO_DNA, roles=[sbol3.SO_PROMOTER])
```

>     >>> ptet.roles = [sbol3.SO_PROMOTER]

Reviewer 2: Set `ptet.roles` and `circuit.roles` in the constructor, as an example of good practice.

Reviewer: ```suggestion
```

>     >>> circuit = sbol3.Component('circuit', sbol3.SBO_DNA)

Reviewer: ```suggestion
    >>> circuit = sbol3.Component('circuit', sbol3.SBO_DNA, roles=[sbol3.SO_ENGINEERED_REGION])
```

>     >>> circuit.roles.append(sbol3.SO_ENGINEERED_REGION)

Reviewer: ```suggestion
```

    
    >>> ptet_sc = sbol3.SubComponent(ptet)
    >>> circuit.features += [ptet_sc]

.. end

-----------------------------------------
Creating and Editing Reference Properties
-----------------------------------------

Some SBOL objects point to other objects by way of URI references. For example, Components point to their corresponding Sequences by way of a URI reference. These kind of properties correspond to white diamond arrows in UML diagrams, as shown in the figure above. Attributes of this type contain the URI of the related object.

.. code:: python

    >>> gfp = sbol3.Component('GFP', sbol3.SBO_DNA)
    >>> gfp_seq = sbol3.Sequence('GFPSequence', elements='atgnnntaa', encoding=sbol3.IUPAC_DNA_ENCODING)
    >>> gfp.sequences = [ gfp_seq ]
    >>> print(gfp.sequences)
    ['http://sbolstandard.org/testfiles/GFPSequence']

.. end

Note that assigning the ``gfp_seq`` object to the ``gfp.sequences`` actually results in assignment of the object's URI. An equivalent assignment is as follows:

.. code:: python

    >>> gfp.sequences = [ gfp_seq.identity ]
    >>> print(gfp.sequences)

>     ['http://sbolstandard.org/testfiles/GFPSequence']

Reviewer 2: Add a example of using lookup() on an element of gfp.sequences

Reviewer: Examples have been added.
